addons/mail/models/mail_channel.py

```diff
@@ -724,30 +726,51 @@ def _execute_channel_pin(self, pinned=False):
+        :param last_message_id: the id of the message to be marked as seen, last message of the
+        thread by default. This param SHOULD be required, the default behaviour is DEPRECATED and
+        kept only for compatibility reasons.
+        """
         self.ensure_one()
-        if self.channel_message_ids.ids:
-            last_message_id = self.channel_message_ids.ids[0] # zero is the index of the last message
-            channel_partner = self.env['mail.channel.partner'].search([('channel_id', 'in', self.ids), ('partner_id', '=', self.env.user.partner_id.id)], limit=1)
-            if channel_partner.seen_message_id.id == last_message_id:
-                # last message seen by user is already up-to-date
+        domain = [('channel_ids', 'in', self.ids)]
+        if last_message_id:
+            domain = expression.AND([domain, [('id', '<=', last_message_id)]])
+        last_message = self.env['mail.message'].search(domain, order="id DESC", limit=1)
+        if not last_message:
+            return
+
+        channel_partner = self.env['mail.channel.partner'].search([
+            ('channel_id', 'in', self.ids),
+            ('partner_id', '=', self.env.user.partner_id.id)
+        ], limit=1)
+        if channel_partner:
+            if channel_partner.seen_message_id and channel_partner.seen_message_id.id >= last_message.id:
                 return
             channel_partner.write({
-                'seen_message_id': last_message_id,
-                'fetched_message_id': last_message_id,
+                'fetched_message_id': last_message.id,
+                'seen_message_id': last_message.id,
             })
-            data = {
-                'id': channel_partner.id,
-                'info': 'channel_seen',
-                'last_message_id': last_message_id,
+        else:
+            channel_partner.create({
+                'channel_id': self.id,
+                'fetched_message_id': last_message.id,
                 'partner_id': self.env.user.partner_id.id,
-            }
-            if self.channel_type == 'chat':
-                self.env['bus.bus'].sendmany([[(self._cr.dbname, 'mail.channel', self.id), data]])
-            else:
-                data['channel_id'] = self.id
-                self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', self.env.user.partner_id.id), data)
-            return last_message_id
+                'seen_message_id': last_message.id,
+            })
+
+        data = {
+            'info': 'channel_seen',
+            'last_message_id': last_message.id,
+            'partner_id': self.env.user.partner_id.id,
+        }
+        if self.channel_type == 'chat':
+            self.env['bus.bus'].sendmany([[(self._cr.dbname, 'mail.channel', self.id), data]])
+        else:
+            data['channel_id'] = self.id
+            self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', self.env.user.partner_id.id), data)
+        return last_message_id
```

Comment:
Github is giving me diff on those lines but I see nothing changed, please double check (whispace maybe?)

Reply:
I think it's because Github consider I changed "data = {" by "else:" so all the rest is a diff... Maybe will consider it correctly when I removed the created